services/gateway/endpoints/src/main/java/org/eclipse/ditto/services/gateway/endpoints/routes/cloudevents/CloudEventsRoute.java

```diff
@@ -0,0 +1,281 @@
+     */
+    private @Nullable
+    CloudEvent toCloudEvent(final RequestContext ctx, final DittoHeaders dittoHeaders, final ByteString payload) {
+
+        if (LOGGER.isTraceEnabled()) {
+            LOGGER.trace("Headers:");
+            ctx.getRequest().getHeaders().forEach(header -> LOGGER.trace("\t{} = {}", header.name(), header.value()));
+            LOGGER.trace("Ditto:");
+            dittoHeaders.forEach((k, v) -> LOGGER.trace("\t{} = {}", k, v));
+        }
+
+        // create a reader for the message
+        final MessageReader reader = HttpMessageFactory.createReader(acceptor -> {
+
+            // NOTE: this acceptor may be run multiple times by the message reader
+
+            // record if we saw the content type header
+            final AtomicBoolean sawContentType = new AtomicBoolean();
+            // consume the HTTP request headers
+            ctx.getRequest().getHeaders().forEach(header -> {
+                if (header.lowercaseName().equals(DittoHeaderDefinition.CONTENT_TYPE.getKey())) {
+                    sawContentType.set(true);
+                }
+                acceptor.accept(header.name(), header.value());
+            });
+
+            if (!sawContentType.get()) {
+                // we didn't see the content type in the header, so extract it from akka's request
+                acceptor.accept(DittoHeaderDefinition.CONTENT_TYPE.getKey(), ctx.getRequest().entity().getContentType().mediaType().toString());
+            }
+        }, payload.toArray());
+
+        return reader.toEvent();
+    }
+
+    private CloudEvent validateCloudEvent(@Nullable final CloudEvent cloudEvent, final RequestContext ctx, final DittoHeaders dittoHeaders) {
+
+        if (cloudEvent == null || cloudEvent.getData() == null) {
+            throw MissingPayloadException
+                    .withDetailedInformationBuilder()
+                    .dittoHeaders(dittoHeaders)
+                    .build();
+        }
+
+        LOGGER.debug("Cloud event: {}", cloudEvent);
+
+        ensureDataContentType(cloudEvent.getDataContentType(), ctx, dittoHeaders);
+        ensureDataSchema(cloudEvent.getDataSchema(), ctx, dittoHeaders);
+
+        return cloudEvent;
+    }
+
+    private Optional<Signal<?>> jsonToDittoSignal(@Nullable final CloudEventData data, final DittoHeaders dittoHeaders) {
+        if (data == null) {
+            return Optional.empty();
+        }
+        final byte[] payload = data.toBytes();
+        if (payload == null || payload.length == 0) {
+            return Optional.empty();
+        }
+
+        JsonObject jsonObject = JsonObject.of(payload);
+        LOGGER.debug("JSON: {}", jsonObject);
+
+        final JsonifiableAdaptable jsonifiableAdaptable = ProtocolFactory.jsonifiableAdaptableFromJson(jsonObject);
+        // use empty header translator to pass along the ditto-auth-context information
+        final DittoProtocolAdapter dittoProtocolAdapter = DittoProtocolAdapter.of(HeaderTranslator.empty());
```

Comment:
Please instantiate the `DittoProtocolAdapter` only once as static field in this class.